Display a deposits paused notice when the account has a negative available balance

haszari opened the pull request:
Fixes #5730 

#### Changes proposed in this Pull Request

This PR introduces a notice shown to merchants when their available balance is negative and may lead to paused deposits.

<img width="708" alt="Screenshot 2023-04-18 at 1 28 57 pm" src="https://user-images.githubusercontent.com/8490476/232663679-3c1c98bc-3e17-423f-9fc3-05e3d36aa56d.png">

#### Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

1. Turn on the new deposit UI: `update_option( '_wcpay_feature_simplify_deposits_ui', 1 );`
2. Purchase any product using a test card that bypasses the pending period (`4000000000000077`)
3. Refund the newly created order. 
4. Go to the Payments > Overview screen.
5. Note that the negative balance notice is shown (see screenshot above). You may need to wait a minute for the transaction to flow through.
6. Make another purchase. 
7. Note that the negative balance notice is no longer displayed. 

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

Commit:
consistently use CardBody container on deposits and balances cards

Commit:
remove placeholder footer, shift footer out of body

Commit:
remove extraneous horizontal padding from recent deposits table:
- now provided by CardBody

Commit:
use CardBody and CardDivider to standardise table styling

Commit:
work in progress / hack in negative deposit notice

Commit:
override unnecessary default notice margins => rely on CardBody padding

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `5957` or branch name `fix/5730-negative-balance-deposits-paused-notice` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&branches.woocommerce-payments=fix/5730-negative-balance-deposits-paused-notice) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: df82feaef6643ca5cc0d702156e3fc4523010fd5
- Build time: 2023-04-18 05:18:35 UTC

_Note: the build is updated when a new commit is pushed to this PR._

haszari committed:
Merge branch 'develop' into fix/use-cardbody-component-for-standard-padding

haszari committed:
Merge branch 'fix/use-cardbody-component-for-standard-padding' into fix/5730-negative-balance-deposits-paused-notice

Commit:
Merge branch 'develop' into fix/5730-negative-balance-deposits-paused-notice:
- with conflicts, manually fixed 🫠

james-allan commented on client/components/deposits-information/index.tsx:
```diff
@@ -106,31 +106,41 @@ const DepositsInformationOverview: React.FunctionComponent< OverviewProps > = (
-					children={ depositButton }
-				/>
-			</Flex>
+			<CardBody>
```
I think this may be a commit from a previous issue you were working on. Is that correct?

haszari replied:
Confirmed, I accidentally hacked on the old deposits overview (2x2 grid) card and this branch was branched from an old copy – removed 😁

james-allan committed:
Fix Div spacing and extra whitespace

james-allan committed:
Rename the amount prop of the next deposit and introduce the raw amount figure

james-allan committed:
Display the negative balance notice when the next deposit is negative

james-allan commented on client/components/deposits-overview/next-deposit.tsx:
```diff
@@ -53,6 +53,22 @@ const NextDepositDetails: React.FC< NextDepositProps > = ( {
+						value={ strings.nextDeposit.description }
+					/>
+				</span>
+			</CardBody>
```
This might also be a merge conflict or copied from another commit. This addition is duplicating the next deposit heading. 

<img width="706" alt="Screenshot 2023-04-13 at 5 16 31 pm" src="https://user-images.githubusercontent.com/8490476/231683074-874f6982-502a-4804-8906-e6a329f82143.png">

haszari replied:
Spot on - this is a merge merge conflict conflict. I've merged this and fixed it twice now (same thing happened on other branch!!)

james-allan commented on client/components/deposits-overview/next-deposit.tsx:
```diff
@@ -140,6 +156,36 @@ const NextDepositDetails: React.FC< NextDepositProps > = ( {
 				) }
+
+				{ /* Notice(s) */ }
+				{ ! isLoading && nextDeposit.amount < 0 && (
```
@haszari I had a crack at using `interpolateComponents` to include the "Why?" link. I've also added the logic so that this notice is shown when the next deposit is negative, however I'm not 100% convinced if that's correct. It's probably supposed to be check if the available balance is negative, but that will need confirmation.

james-allan commented on client/components/deposits-overview/utils.ts:
```diff
@@ -38,6 +40,7 @@ export const getNextDeposit = (
 		status: nextScheduled.status ?? 'estimated',
-		amount: formatCurrency( nextScheduled.amount ?? 0, currency ),
+		formattedAmount: formatCurrency( nextScheduled.amount ?? 0, currency ),
+		amount: nextScheduled.amount ?? 0,
```
I had to introduce an raw `amount` figure so I could use the figure in logic without having to reverse format it. 

If it turns out we're not suppose to use the next deposit amount this can be reverted.

haszari replied:
I _think_ we can revert this but need to get the logic right first!

james-allan reviewed (changes requested):
@haszari I've submitted a couple of changes to:

1. Introduce logic to only show the notice when the next deposit is negative *
2. `interpolateComponents` the notice and add the URL. 

\* as I mentioned in one of the review comments, I'm not 100% sure what the source of the check should be. A negative available balance or a negative next deposit.

haszari commented:
Thanks for the assist @james-allan ! Looks like my mega merge went a bit awry. I'll continue on this today :)

james-allan commented:
@haszari just a heads up that I've merged https://github.com/Automattic/woocommerce-payments/pull/6022 which will cause a conflict with this branch. The change required on this branch should be to just move the notice inside the Notices section and remove any unnecessary css changes. 

Let me know if you'd like a hand resolving it.

Commit:
Merge branch 'develop' into fix/5730-negative-balance-deposits-paused-notice:
- manually fixed conflicts in next-deposit.tsx

Commit:
revert use of `CardBody` in legacy deposits overview card

Commit:
remove duplicate "next deposit" heading – fix bad merge

james-allan commented:
> as I mentioned in one of the review comments, I'm not 100% sure what the source of the check should be. A negative available balance or a negative next deposit.

I think we have an answer on this now. It is supposed to be the **available balance** that forms the condition on whether this notice is shown, not the next deposit, like I submitted.

haszari commented:
> > as I mentioned in one of the review comments, I'm not 100% sure what the source of the check should be. A negative available balance or a negative next deposit.

> I think we have an answer on this now. It is supposed to be the available balance that forms the condition on whether this notice is shown, not the next deposit, like I submitted.

Thanks for confirming @james-allan - that makes sense to me.

- If your WCPay balance is negative, you won't get any payouts. "Deposits paused" until you're in credit.
- If deposits are paused, the "next deposit" makes no sense/doesn't exist.

In addition I think we should show `$0` as in the designs. If this isn't happening already I'll include this in the current PR.

Commit:
add negative balance component and show in UI (hard-coded for now) +
+ refactor all next deposit notices into named components

Commit:
move strings inline so components are readable and self-contained

Commit:
restore link for waiting period notice (pasteo in previous commit)

Commit:
restore logic for showing waiting period notice (fix js tests)

Commit:
update snapshot (1x test still failing, will revisit later)

Commit:
only show paused notice if balance (next deposit) is negative:
- TBC, this may be incorrect logic, other balances are not negative in my test store

haszari commented on client/components/deposits-overview/next-deposit.tsx:
```diff
@@ -49,12 +130,14 @@ const NextDepositDetails: React.FC< NextDepositProps > = ( {
 	);
 	const completedWaitingPeriod =
 		wcpaySettings.accountStatus.deposits?.completed_waiting_period;
+	const negativeBalanceDepositsPaused = overview?.nextScheduled?.amount
```
Unclear if this is the correct balance to check. I'm using this for now because this is the negative value in my store right now, but we need to confirm this and ensure we test a variety of scenarios.

haszari commented on client/components/deposits-overview/next-deposit.tsx:
```diff
@@ -31,6 +30,88 @@ type NextDepositProps = {
+		} ) }
+	</BannerNotice>
+);
+
```
I factored these out to keep the main component slim, reduce indentation, and make the show/hide logic easy to read.

haszari commented on client/components/deposits-overview/next-deposit.tsx:
```diff
@@ -64,19 +147,22 @@ const NextDepositDetails: React.FC< NextDepositProps > = ( {
 						<Loadable
 							isLoading={ isLoading }
-							value={ strings.tableHeaders.amount }
+							value={ __( 'Amount', 'woocommerce-payments' ) }
```
I also went on a tear moving strings inline. This might have been a bit ambitious – I could be convinced to back out 😁

haszari commented on client/components/deposits-overview/utils.ts:
```diff
@@ -7,7 +7,8 @@ type NextDepositTableData = {
 	status: string;
-	amount: string;
+	formattedAmount: string;
+	amount: number;
```
We might not need this any more – will need to remember to revert when everything else is working correctly.

james-allan replied:
Yeah this isn't needed anymore. I've reverted it in 83a9afa615114f2a3183217e1462e002963df5ca

haszari commented on client/overview/style.scss:
```diff
@@ -28,7 +28,7 @@
 
 .wcpay-overview {
-	.components-notice {
+	> .components-notice {
```
This may be unnecessary, will recheck when show/hide logic is working well.

haszari commented:
If anyone wants to resume work on this, feel free. 

- Notice wording is as per design and now includes docs link.
- Notice is supposed to hide if next deposit is not negative (TBC). I have not successfully tested this …
- I've factored out the notices so main component is leaner and logic easy to follow.
- I've moved various strings inline (motivated by the notice component refactor above, so they are readable and relatively self-contained).

Next steps:

- Get the notice showing when appropriate. Is supposed to be when (selected currency) balance is negative, but this is $0 for me (when next deposit is negative). Might need changes upstream (data / server). 
- Fix tests (`Confirm new account waiting period notice shows` is failing)
- Review and remove unnecessary / experimental changes from earlier iterations
- Review => merge

james-allan committed:
Use the currently selected currency's available balance to determine of the notice is shown

james-allan commented:
@haszari I've pushed [179bb93](https://github.com/Automattic/woocommerce-payments/pull/5957/commits/179bb938ab7d675c0291bf468121de46d11cca80) which includes the changes to use the currently selected currency available balance as the check for this notice being shown. 

**Positive balance:** 
<img width="591" alt="Screenshot 2023-04-14 at 3 20 48 pm" src="https://user-images.githubusercontent.com/8490476/231948311-f1666c0c-0318-4ab7-9ef1-a29080ba6ca3.png">

**Negative balance:** 
<img width="591" alt="Screenshot 2023-04-14 at 3 33 59 pm" src="https://user-images.githubusercontent.com/8490476/231950188-83ce15e7-c6b9-457b-a7fe-6c2d33cec750.png">

### Next
- Test this on a JN site with multiple currencies.
- Review and remove unnecessary / experimental changes from earlier iterations

james-allan committed:
Tighten logic so it only falls through to default case if a selected currency isn't chosen

james-allan commented:
I just tested on a JN site and confirmed this works as intended. 

https://user-images.githubusercontent.com/8490476/231967732-93cb11a9-8924-4d44-8472-7ec4a5c9240d.mov

james-allan committed:
Revert "Rename the amount prop of the next deposit and introduce the raw amount figure"

This reverts commit f956c9279ef58cb4d78923ab5a99090067a17a05.

james-allan committed:
Merge branch 'develop' into fix/5730-negative-balance-deposits-paused-notice

james-allan committed:
Merge branch 'develop' into fix/5730-negative-balance-deposits-paused-notice

james-allan committed:
Add changelog entry

james-allan commented:
I've written the PR description and pulled the latest set of changes. This still works for me. And looks correct in the various forms. 

|     No history | With history |
| ----------- | ----------- |
| <img width="703" alt="Screenshot 2023-04-18 at 1 56 11 pm" src="https://user-images.githubusercontent.com/8490476/232667419-d4199c4c-fc40-4e5f-8a7f-c2030846b866.png">      | <img width="702" alt="Screenshot 2023-04-18 at 1 56 42 pm" src="https://user-images.githubusercontent.com/8490476/232667391-72a9c7d2-bfd3-4db9-bd5c-0f735983b29f.png">|

james-allan committed:
Fix unit tests

james-allan committed:
Add 2 tests to confirm the negative balance notice is and isn't shown under expected circumstances

james-allan commented:
I'm going to merge this one on Rua's behalf so he can focus on other team priorities.
